src/hypothesis/extra/django/models.py

```diff
@@ -17,159 +17,402 @@
 from django.core.exceptions import ValidationError
 
-import hypothesis.strategies as st
-import hypothesis.extra.fakefactory as ff
+from hypothesis import strategies as st
 from hypothesis.errors import InvalidArgument
-from hypothesis.extra.datetime import datetimes
-from hypothesis.utils.conventions import UniqueIdentifier
+from hypothesis.extra.datetime import dates, datetimes, times
+from hypothesis.extra.fakefactory import fake_factory
 from hypothesis.searchstrategy.strategies import SearchStrategy
+from hypothesis.strategies import defines_strategy
 
 
-class ModelNotSupported(Exception):
-    pass
+@defines_strategy
+def model_text(alphabet=st.characters(blacklist_characters='\x00',
+                                      blacklist_categories=('Cs',)), **kwargs):
+    """A modified text strategy that plays nicer with Django databases.
+
+    - Excludes the null character, which results in field truncation
+      in Postgres.
+
+    """
+    return st.text(alphabet=alphabet, **kwargs)
+
+
+# Field strategies.
+
+_field_mappings = {}
+
+
+def add_default_field_mapping(field_type, strategy=None):
+    """Registers a strategy or strategy factory as the default handler for a
+    Django model field type."""
+    # Allow use as a decorator.
+    if strategy is None:
+        return partial(add_default_field_mapping, field_type)
+    # Allow use directly.
+    _field_mappings[field_type] = strategy
+    # Return the strategy, to complete usage as a decorator.
+    return strategy
+
+
+@add_default_field_mapping(dm.AutoField)
+@defines_strategy
+def default_value(field):
+    """A strategy that returns the default value for the field."""
+    return st.just(field.get_default())
+
+
+def validator_to_filter(field):
+    """Creates a filter for the given field that filters out values that do not
+    pass field validation."""
+    def validate(value):
+        try:
+            field.run_validators(value)
+            return True
+        except ValidationError:
+            return False
+    return validate
+
+
+def defines_field_strategy(blank_values=()):
+    """
+    Decorates the given field strategy factory to add support for:
+
+    - Fields with choices.
+    - Null fields.
+    - Filtering by field validators.
+
+    """
+    def decorator(func):
+        @defines_strategy
+        @wraps(func)
+        def create_field_strategy(field, **kwargs):
+            # Handle field choices.
+            if field.choices:
+                choices = [
+                    value
+                    for (value, name)
+                    in field.choices
+                ]
+                if field.blank:
+                    choices.extend(blank_values)
+                strategy = st.sampled_from(choices)
+            else:
+                # Run the factory function.
+                strategy = func(field, **kwargs)
+            # Add in null values.
+            if field.null:
+                strategy = st.one_of(st.none(), strategy)
```

Comment:
In cases where the field is nullable but has a `choices` attribute, this will strongly bias towards null values. Would it make sense to add `None` to the choices list passed to `sampled_from`?